syntax = "proto3";

package genx;

service ModelContextService {
  rpc LoadModel(LoadModelRequest) returns (LoadModelResponse) {}
  rpc UnloadModel(UnloadModelRequest) returns (UnloadModelResponse) {}
  rpc GetModelContext(GetModelContextRequest) returns (GetModelContextResponse) {}
  rpc GenerateEmbedding(GenerateEmbeddingRequest) returns (GenerateEmbeddingResponse) {}
}

message LoadModelRequest {
  string model_id = 1;
  string model_name = 2;
  string model_type = 3; // "embedding", "llm", etc.
  map<string, string> config = 4; // e.g., {"device": "cuda"}
}

message LoadModelResponse {
  string model_id = 1;
  string status = 2;
  string error = 3;
}

message UnloadModelRequest {
  string model_id = 1;
}

message UnloadModelResponse {
  string model_id = 1;
  string status = 2;
  string error = 3;
}

message GetModelContextRequest {
  string model_id = 1;
}

message GetModelContextResponse {
  string model_id = 1;
  string model_name = 2;
  string model_type = 3;
  string status = 4;
  map<string, string> config = 5;
  string error = 6;
}

message GenerateEmbeddingRequest {
  string model_id = 1;
  repeated string texts = 2;
}

message GenerateEmbeddingResponse {
  string model_id = 1;
  string status = 2;
  repeated float embeddings = 3;
  string error = 4;
}